app/backend/src/couchers/servicers/requests.py

```diff
@@ -584,3 +593,81 @@ def MarkLastSeenHostRequest(self, request, context):
 
             session.commit()
             return empty_pb2.Empty()
+
+    def GetResponseRate(self, request, context):
+        with session_scope() as session:
+            # this subquery gets the time that the request was sent
+            t = (
+                select(Message.conversation_id, Message.time)
+                .where(Message.message_type == MessageType.chat_created)
+                .subquery()
+            )
+            # this subquery gets the time that the user responded to the request
+            s = (
+                select(Message.conversation_id, func.min(Message.time).label("time"))
+                .where(Message.author_id == request.user_id)
+                .group_by(Message.conversation_id)
+                .subquery()
+            )
+
+            res = session.execute(
+                select(
+                    User.id,
+                    # number of requests received
+                    func.count().label("n"),
+                    # percentage of requests responded to
+                    (func.count(s.c.time) / func.cast(func.greatest(func.count(t.c.time), 1.0), Float)).label(
+                        "response_rate"
+                    ),
+                    # the 33rd percentile response time
+                    percentile_disc(0.33)
+                    .within_group(func.coalesce(s.c.time - t.c.time, timedelta(days=1000)))
```

Comment:
The coalesce function returns the first of its arguments that is not null, but how can the result of a subtraction be null? I don't understand, to me it sounds like we are excluding messages that haven't been responded to.

Reply:
It's an outer join, so if there is no response to a message (no corresponding row in the `s` subquery) then `s.c.time - t.c.time` will be null.

Reply:
Thanks for explaining SQL to me :)